### DISCUSSION TITLE
Welcome to OpenDiS Discussions!

### DISCUSSION THREAD

**Discussion by nrbertin** [OpenDiS/OpenDiS#7] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Passion28-y** [comment]


**Reply by nrbertin** [reply]
I strongly encourage new users to "learn" OpenDiS instead of ParaDiS. Although ParaDiS will continue to be available as a legacy code, we no longer support it.

OpenDiS is a fully open-source project based on new key features that should render the code much easier to use and more efficient than ParaDiS was. These [key features](https://opendis.github.io/OpenDiS/manifesto.html) include a python-based interface, a modular design, and a GPU-accelerated implementation.

**Comment by Passion28-y** [comment]
I want to know how to define the initial dislocation configuration intuitively.

**Reply by shaswat-m** [reply]
You can check this thread: https://github.com/OpenDiS/OpenDiS/issues/5

**Reply by Passion28-y** [reply]
thank you

**Reply by Passion28-y** [reply]
@shaswat-m But I'm using ParaDiS now, and the operation seems different.

**Reply by nrbertin** [reply]
ParaDiS uses .data input files for the initial configuration, which can be generated using OpenDiS. **Note however that I will strongly encourage new users to "learn" OpenDiS instead of ParaDiS. Although ParaDiS will continue to be available as a legacy code, we no longer support it.**

Perhaps for a more intuitive start you can experiment with generating simple configurations, e.g. a single Frank-Read source, using utility functions provided in OpenDiS. For instance, here is a script to generate a single 1/2\[110\](-111) edge Frank-Read source of length 500b discretized into 10 nodes in a box of size 1000b:

```
import pyexadis
from pyexadis_base import ExaDisNet
from pyexadis_utils import insert_frank_read_src
pyexadis.initialize()

# Define box
Lbox = 1000.0
cell = pyexadis.Cell(Lbox)
nodes, segs = [], []

# Insert Frank-Read source
burg = 1.0/np.sqrt(2.0)*np.array([1.,1.,0.])
plane = np.array([-1.,1.,1.])
length = 0.5*Lbox
center = cell.center()
nodes, segs = insert_frank_read_src(cell, nodes, segs, burg, plane,
                                    length, center, theta=90.0, numnodes=10)
# Export in .data file
G = ExaDisNet(cell, nodes, segs)
G.write_data('frs.data')
```
Running this script will generate file `frs.data`, e.g. which can then be used as an input configuration for ParaDiS or OpenDiS. Additional utility functions to generate different initial configurations (e.g. dislocation loops, infinite lines and dipoles, etc.) are documented [here](https://opendis.github.io/OpenDiS/core_libraries/exadis_documentation/user_guide/python_modules.html#dislocation-network).

**Reply by Passion28-y** [reply]
@nrbertin Ok, but I have been studying ParaDiS for some time, but there are still some visualization tools that I can't use. Maybe I can learn about using it together.Thank you for your advice.

**Reply by Passion28-y** [reply]
@nrbertin Dear developers, I have a question, if I want to set an initial dislocation configuration for the FCC, can I use the Frank-Read source to automatically generate the relevant data files? Is it true that most of the time the data file is generated by inserting into the Frank-Read source?
One more question, I don't know when this Frank-Read source will stop proliferating?
<img width="556" alt="截屏2024-12-07 18 58 07" src="https://github.com/user-attachments/assets/0a686e02-0e38-43ef-965d-f5acc85d6f65">

**Reply by nrbertin** [reply]
Frank-Read sources are a good way to start and generate toy configurations, or configurations for specific applications (e.g. binary junction formation, etc.). However, Frank-Read sources made of a single dislocation line terminating in the crystal at both ends are rather unphysical objects that introduce artificial pinning points. Thus, to avoid these artifacts, in bulk DDD simulations (e.g. work-hardening simulations) we generally use infinite lines to generate the initial configurations. To generate infinite lines, you can look at this issue #5.

**Reply by Passion28-y** [reply]
@nrbertin Got it, thank you very much. I have another question about ParaDiS, I hope you can help me answer it. Regarding the selection of <timestepIntegrator>, I saw that Professor Cai Wei used a method called sub-loop in his paper, but there is no such option in ParaDiS. Do I need to modify the code myself? Looking forward to your answer.

**Reply by nrbertin** [reply]
@Passion28-y Not sure what you are referring to. If you point me to the specific paper then I may be able to help.

**Reply by Passion28-y** [reply]
@nrbertin OK, the title of the paper is Prediction of Yield Surface of Single Crystal Copper from Discrete Dislocation Dynamics and Geometric Learning. Thank you very much.

**Reply by Passion28-y** [reply]
![IMG_20241210_100437.jpg](https://github.com/user-attachments/assets/2dc1152e-f4a4-4685-8c3b-3e648f8e572b)

**Reply by nrbertin** [reply]
@Passion28-y I'm assuming you are referring to the subcycling approach then. It is not available in the LLNL version of the code, but is available in its [Stanford branch](https://gitlab.com/opendis/ParaDiS-2.7). It is also available in OpenDiS.

**Reply by Passion28-y** [reply]
@nrbertin Yes, your assumption is correct. Thank you very much for your help. I should learn OpenDiS carefully.

**Reply by Passion28-y** [reply]
@nrbertin Excuse me, what software can be used for visualization and post-processing of the output files after simulation?

**Reply by nrbertin** [reply]
@Passion28-y I think the only visualization tool that natively supports .data files is [VisIt](https://visit.llnl.gov/). If you have Ovito Pro then you can look at this thread #3. You can also convert .data files to .vtk files for visualization in a software like ParaView, see #4.

**Reply by Passion28-y** [reply]
@nrbertin OK, I understand. Thank you very much.

**Reply by Passion28-y** [reply]
@nrbertin I have a question, is it possible to add precipitates or voids when building the initial model using OpenDiS? I have not found any relevant information, and I look forward to your help.

**Reply by nrbertin** [reply]
@Passion28-y Right now we do not have a module to explicitly handle precipitates or voids. This is something we plan to support in the future.

**Reply by Passion28-y** [reply]
@nrbertin OK, thank you for your answer

**Reply by Passion28-y** [reply]
@nrbertin I see that ParaDiS can add precipitates, but I don't see whether it can add voids?

**Reply by nrbertin** [reply]
@Passion28-y No, voids are not implemented.

**Reply by Passion28-y** [reply]
@nrbertin Thank you very much

**Reply by Passion28-y** [reply]
@nrbertin In ParaDiS, I can generate a precipitate file, but I cannot call it in the ctrl file. In addition, when generating the precipitate, can't I specify the size and position of the precipitate? I have been thinking about this for a long time. I hope you can answer my confusion. Thank you very much.

**Reply by nrbertin** [reply]
@Passion28-y For clarity, please repost your question on the ParaDiS github.

**Reply by Passion28-y** [reply]
OK,thank you

**Comment by Essea** [comment]
Dear Developers, am I right that FCC_climb (on the analogy of Paradis) has not been implemented here yet? Do you consider current list of mobility laws for FCC crystals to be extended in the future? Thank you.

**Reply by nrbertin** [reply]
@Essea The FCC_climb mobility is not implemented, but mobility `FCC_0B` available from pyexadis should be relatively equivalent as it allows for a climb component of the velocity, see documentation [here](https://opendis.github.io/OpenDiS/core_libraries/exadis_documentation/user_guide/python_modules.html#mobility-laws). To allow for climb motion, option `state["enforce_glide_planes"] = 0` must be specified, otherwise node velocities will be systematically projected back to glide planes.

Note: as for FCC_climb, climb in `FCC_0B` is modeled in a glide-like manner via a linear climb mobility coefficient. This is a rather crude representation of climb, which physically is a diffusion-controlled process.
